operator/src/lib.rs

```diff
@@ -131,6 +135,187 @@ fn get_pods_for_node(hashed: &HashMap<String, Vec<Pod>>) -> Vec<Pod> {
 }
 
 impl SparkState {
+    async fn set_upgrading_condition(
+        &self,
+        conditions: &[Condition],
+        message: &str,
+        reason: &str,
+        status: ConditionStatus,
+    ) -> OperatorResult<SparkCluster> {
+        let resource = self
+            .context
+            .build_and_set_condition(
+                Some(conditions),
+                message.to_string(),
+                reason.to_string(),
+                status,
+                "Upgrading".to_string(),
+            )
+            .await?;
+
+        Ok(resource)
+    }
+
+    async fn set_current_version(
+        &self,
+        version: Option<&SparkVersion>,
+    ) -> OperatorResult<SparkCluster> {
+        let resource = self
+            .context
+            .client
+            .merge_patch_status(
+                &self.context.resource,
+                &json!({ "currentVersion": version }),
+            )
+            .await?;
+
+        Ok(resource)
+    }
+
+    async fn set_target_version(
+        &self,
+        version: Option<&SparkVersion>,
+    ) -> OperatorResult<SparkCluster> {
+        let resource = self
+            .context
+            .client
+            .merge_patch_status(&self.context.resource, &json!({ "targetVersion": version }))
+            .await?;
+
+        Ok(resource)
+    }
+
+    /// Will initialize the status object if it's never been set.
+    pub async fn init_status(&mut self) -> SparkReconcileResult {
+        // We'll begin by setting an empty status here because later in this method we might
+        // update its conditions. To avoid any issues we'll just create it once here.
+        if self.status.is_none() {
+            let status = SparkClusterStatus::default();
+            self.context
+                .client
+                .merge_patch_status(&self.context.resource, &status)
+                .await?;
+            self.status = Some(status);
+        }
+
+        // This should always return either the existing one or the one we just created above.
+        let status = self.status.take().unwrap_or_default();
+        let spec_version = self.spec.version.clone();
+
+        // TODO: This is not ideal: We pass in status.conditions to `set_upgrading_conditions` which
```

Comment:
I think we're handlnig this now because we have the return value of `set_upgrading_condition`